pkg/controller/autoscaling/elasticsearch/autoscaler/offline.go

```diff
@@ -0,0 +1,129 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package autoscaler
+
+import (
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/autoscaling/elasticsearch/resources"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/autoscaling/elasticsearch/status"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// GetOfflineNodeSetsResources attempts to create or restore resources.NodeSetsResources without an actual autoscaling
+// decision from Elasticsearch. It ensures that even if no decision has been returned by the autoscaling API then
+// the NodeSets still respect the min. and max. resource requirements specified by the user.
+// If resources are within the min. and max. boundaries then they are left untouched.
+func GetOfflineNodeSetsResources(
+	log logr.Logger,
+	nodeSets []string,
+	autoscalingSpec esv1.AutoscalingPolicySpec,
+	currentAutoscalingStatus status.Status,
+) resources.NodeSetsResources {
+	currentNodeSetsResources, hasNodeSetsResources := currentAutoscalingStatus.CurrentResourcesForPolicy(autoscalingSpec.Name)
+
+	var nodeSetsResources resources.NodeSetsResources
+	var expectedNodeCount int32
+	if !hasNodeSetsResources {
+		// There's no current status for this nodeSet, this happens when the Elasticsearch cluster does not exist.
+		// In that case we create a new one from the minimum values provided by the user.
+		nodeSetsResources = newMinNodeSetResources(autoscalingSpec, nodeSets)
+	} else {
+		// The status contains some resource values for the NodeSets managed by this autoscaling policy, let's reuse them.
+		nodeSetsResources = nodeSetResourcesFromStatus(currentAutoscalingStatus, currentNodeSetsResources, autoscalingSpec, nodeSets)
+		for _, nodeSet := range currentNodeSetsResources.NodeSetNodeCount {
+			expectedNodeCount += nodeSet.NodeCount
+		}
+	}
+
+	// Ensure that the min. number of nodes is in the allowed range.
+	if expectedNodeCount < autoscalingSpec.NodeCount.Min {
+		expectedNodeCount = autoscalingSpec.NodeCount.Min
+	} else if expectedNodeCount > autoscalingSpec.NodeCount.Max {
+		expectedNodeCount = autoscalingSpec.NodeCount.Max
+	}
+
+	// User may have added or removed some NodeSets while the autoscaling API is not available.
+	// We distribute the nodes to reflect that change.
+	fnm := NewFairNodesManager(log, nodeSetsResources.NodeSetNodeCount)
+	for expectedNodeCount > 0 {
+		fnm.AddNode()
+		expectedNodeCount--
+	}
```

Comment:
This abstraction strikes me as a bit odd. Could this not simply be a function?
```golang
distributeFairly(nodeSetsResources, expectedNodeCount)
```

Also: the logger seems unused.

Reply:
You're right, done in 3f462982f